lenskit-core/src/main/java/org/grouplens/lenskit/data/event/Like.java

```diff
@@ -20,14 +20,120 @@
  */
 package org.grouplens.lenskit.data.event;
 
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+import java.io.Serializable;
+
 /**
  * Generic interface representing a unary interaction.  This provides a basic, generic representation
  * of likes, clicks, purchases, and other unary interactions.  Applications may want to implement
  * their own events to represent such interactions.
+ * <p>
+ * This class can be created with {@link LikeBuilder} or the {@link #create(long, long, long)} method.
+ * </p>
  *
  * @since 2.2
  * @see Events#like(long, long)
  * @see Events#like(long, long, long)
  */
-public interface Like extends Event {
+public final class Like implements Event, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final long userId;
+    private final long itemId;
+    private final long timestamp;
+
+    Like(long uid, long iid, long ts) {
```

Comment:
Is it too confusing to have factory methods, and public constructors, and builders? That would be a defensible design, if perhaps a bit confusing.

Builders have the nice advantage that when you are reading code, you don't have to remember what order things go in, because everything has a name (like Python keyword arguments). For simple events, that isn't a big deal, especially if the code building the object has sensible variable names. But with more than about 4-5 parameters, it can be a bigger deal.

Agree that consistency is already gone with the DI, and that is a decent argument for not trying for a scrupulous consistency here.

I don't have an exhaustive list of factories, and the list is changing anyway because I am starting to use the pattern in more code. It's worth noting that we *don't* have a factory for `LenskitConfiguration`, but do for `LenskitRecommenderEngineBuilder` and `*Loader`. There the difference might be defensible as 'LenskitConfiguration is a "data" object, and the engine builder/loader is a Big Object'. But that seems a bit silly.

Also worth noting: whichever way we go, we are not locked in until we cut a release candidate for LensKit 3.0. We can change all we want before then.